xdsl/dialects/experimental/fir.py

```diff
@@ -0,0 +1,2137 @@
+        # yet
+        parser.parse_char("<")
+        has_tuple = parser.parse_optional_keyword("tuple")
+        if has_tuple is None:
+            typ = parser.parse_type()
+            parser.parse_char(">")
+            return [typ]
+        else:
+            # If its a tuple then there are any number of types
+            def parse_types():
+                return parser.parse_type()
+
+            values = parser.parse_comma_separated_list(
+                parser.Delimiter.ANGLE, parse_types
+            )
+            parser.parse_char(">")
+            return [TupleType(values)]
+
+
+@irdl_attr_definition
+class DeferredAttr(ParametrizedAttribute, TypeAttribute):
+    """
+    A deferred size which is represented with a question mark
+    """
+
+    name = "fir.deferred"
+
+    def print_parameters(self, printer: Printer) -> None:
+        printer.print_string("?")
+
+
+@irdl_attr_definition
+class LLVMPointerType(ParametrizedAttribute, TypeAttribute):
+    """
+    A pointer type that does not have any of the constraints and semantics
+    of other FIR pointer types and that translates to llvm pointer types.
+    It is meant to implement indirection that cannot be expressed directly
+    in Fortran, but are needed to implement some Fortran features (e.g,
+    double indirections).
+    """
+
+    name = "fir.llvm_ptr"
+
+    type: ParameterDef[AnyIntegerAttr | AnyFloat]
+
+
+@irdl_attr_definition
+class NoneType(ParametrizedAttribute, TypeAttribute):
+    """
+    This isn't part of the FIR MLIR dialect, and is only represented internally with
+    xDSL, but is useful to denote when there is a none or empty attribute
+    """
+
+    name = "fir.none"
+
+
+@irdl_attr_definition
+class SequenceType(ParametrizedAttribute, TypeAttribute):
+    """
+    A sequence type is a multi-dimensional array of values. The sequence type
+    may have an unknown number of dimensions or the extent of dimensions may be
+    unknown. A sequence type models a Fortran array entity, giving it a type in
+    FIR. A sequence type is assumed to be stored in a column-major order, which
+    differs from LLVM IR and other dialects of MLIR.
+    """
+
+    name = "fir.array"
+    shape: ParameterDef[ArrayAttr[AnyIntegerAttr | DeferredAttr | NoneType]]
+    type: ParameterDef[IntegerType | AnyFloat | ReferenceType]
+    type2: ParameterDef[IntegerType | AnyFloat | ReferenceType | NoneType]
+
+    def __init__(
+        self,
+        type1: ParameterDef[IntegerType | AnyFloat | ReferenceType],
+        shape: Optional[List[int | IntegerAttr[IndexType] | DeferredAttr]] = None,
+        type2: Optional[ParameterDef[IntegerType | AnyFloat | ReferenceType]] = None,
+    ):
+        if type2 is not None:
+            super().__init__([ArrayAttr([NoneType()]), type1, type2])
+        else:
+            if shape is None:
+                shape = [1]
+            super().__init__(
+                [
+                    ArrayAttr(
```

Comment:
seems like this would be a bit easier to read one line above the init call, will get squished a bit less by black also

Reply:
Thanks, good suggestion - done :)